File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/outsleeping/service/OutSleepingService.java
package b1nd.dodam.domain.rds.outsleeping.service;

import b1nd.dodam.domain.rds.member.entity.Student;
import b1nd.dodam.domain.rds.outsleeping.entity.OutSleeping;
import b1nd.dodam.domain.rds.outsleeping.exception.OutSleepingNotFoundException;
import b1nd.dodam.domain.rds.outsleeping.repository.OutSleepingRepository;
import b1nd.dodam.domain.rds.support.enumeration.ApprovalStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OutSleepingService {

    private final OutSleepingRepository repository;

    public void save(OutSleeping outSleeping) {
        repository.save(outSleeping);
    }

    public void delete(OutSleeping outSleeping) {
        repository.delete(outSleeping);
    }

    public OutSleeping getById(Long id) {
        return repository.findById(id)
                .orElseThrow(OutSleepingNotFoundException::new);
    }

    public List<OutSleeping> getValid(LocalDate now) {
        return getByDate(now).parallelStream().toList();
    }

    public List<OutSleeping> getResidual(LocalDate now){
//        return getByDate(now).parallelStream()
        return null;
    }

    public List<OutSleeping> getByDate(LocalDate date) {
        return repository.findByDate(date);
    }

    public List<OutSleeping> getByEndAt(LocalDate endAt){
        return repository.findByEndAtGreaterThanEqual(endAt);
    }

    public List<OutSleeping> getByStudent(Student student, LocalDate now) {
        return repository.findByStudentAndEndAtGreaterThanEqual(student, now);
    }

}

File: services/service-outsleeping/src/main/kotlin/com/b1nd/dodamdodam/outsleeping/domain/outsleeping/service/OutSleepingService.kt
package com.b1nd.dodamdodam.outsleeping.domain.outsleeping.service

import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.entity.OutSleepingEntity
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.enumeration.OutSleepingStatusType
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingAlreadyExistsException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingForbiddenException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingNotFoundException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingNotPendingException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.repository.OutSleepingRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.UUID

@Service
class OutSleepingService(
    private val repository: OutSleepingRepository
) {
    fun create(entity: OutSleepingEntity): OutSleepingEntity {
        if (repository.existsOverlapping(entity.studentId, entity.startAt, entity.endAt)) {
            throw OutSleepingAlreadyExistsException()
        }
        return repository.save(entity)
    }

    fun findById(id: Long): OutSleepingEntity =
        repository.findById(id).orElseThrow { OutSleepingNotFoundException() }

    fun allow(id: Long) {
        val entity = findById(id)
        entity.status = OutSleepingStatusType.ALLOWED
    }

    fun reject(id: Long, rejectReason: String?) {
        val entity = findById(id)
        entity.status = OutSleepingStatusType.REJECTED
        entity.rejectReason = rejectReason
    }

    fun revert(id: Long) {
        val entity = findById(id)
        entity.status = OutSleepingStatusType.PENDING
        entity.rejectReason = null
    }

    fun delete(id: Long, studentId: UUID) {
        val entity = findById(id)
        if (entity.studentId != studentId) throw OutSleepingForbiddenException()
        if (entity.status != OutSleepingStatusType.PENDING) throw OutSleepingNotPendingException()
        repository.delete(entity)
    }

    fun findByDate(date: LocalDate): List<OutSleepingEntity> =
        repository.findByStartAtLessThanEqualAndEndAtGreaterThanEqual(date, date)

    fun findByStudentId(studentId: UUID): List<OutSleepingEntity> =
        repository.findByStudentId(studentId)

    fun findValid(today: LocalDate): List<OutSleepingEntity> =
        repository.findByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
            OutSleepingStatusType.ALLOWED, today, today
        )
}
